nexus/mgs-updates/src/test_util/updates.rs

```diff
@@ -0,0 +1,376 @@
+use tufaceous_artifact::ArtifactVersion;
+use uuid::Uuid;
+
+/// Describes an update operation that can later be executed any number of times
+pub struct UpdateDescription<'a> {
+    // Execution information
+    pub gwtestctx: &'a GatewayTestContext,
+    pub artifacts: &'a TestArtifacts,
+
+    // Update parameters
+    pub sp_type: SpType,
+    pub slot_id: u32,
+    pub artifact_hash: &'a ArtifactHash,
+
+    // Overrides
+    //
+    // If `None`, the correct value is determined automatically.  These are
+    // overridable in order to induce specific kinds of failures.
+    pub override_baseboard_id: Option<BaseboardId>,
+    pub override_expected_active: Option<ArtifactVersion>,
+    pub override_expected_inactive: Option<ExpectedVersion>,
+    pub override_progress_timeout: Option<Duration>,
+}
+
+impl UpdateDescription<'_> {
+    /// Sets up a single attempt to execute the update described by `self`
+    ///
+    /// Execution does not start until you call `run_until_status()` or
+    /// `finish()`on the returned value.
+    pub async fn setup(&self) -> InProgressAttempt {
+        let mgs_client = self.gwtestctx.client();
+
+        // Fetch information about the device that we're going to update.
+        // This will be used to configure the preconditions (expected baseboard
+        // id and expected active/inactive slot contents).
+        let sp1 = SpTestState::load(&mgs_client, self.sp_type, self.slot_id)
+            .await
+            .expect("loading initial state");
+        let baseboard_id = Arc::new(
+            self.override_baseboard_id
+                .clone()
+                .unwrap_or_else(|| sp1.baseboard_id()),
+        );
+        let expected_active_version = self
+            .override_expected_active
+            .clone()
+            .unwrap_or_else(|| sp1.expect_sp_active_version());
+        let expected_inactive_version = self
+            .override_expected_inactive
+            .clone()
+            .unwrap_or_else(|| sp1.expect_sp_inactive_version());
+        let deployed_caboose = self
+            .artifacts
+            .deployed_caboose(self.artifact_hash)
+            .expect("caboose for generated artifact");
+
+        // Assemble the driver-level update request.
+        let mgs_backends = self.gwtestctx.mgs_backends();
+        let sp_update_request = PendingMgsUpdate {
+            baseboard_id: baseboard_id.clone(),
+            sp_type: self.sp_type,
+            slot_id: self.slot_id,
+            details: PendingMgsUpdateDetails::Sp {
+                expected_active_version,
+                expected_inactive_version,
+            },
+            artifact_hash: *self.artifact_hash,
+            artifact_version: std::str::from_utf8(
+                deployed_caboose.version().unwrap(),
+            )
+            .unwrap()
+            .parse()
+            .unwrap(),
+        };
+
+        let request = sp_update_request.clone();
+        let update_id = Uuid::new_v4();
```

Comment:
Do we want a strongly-typed `SpUpdateUuid`? (It'd have to be squished down to `Uuid` when handed off to crates from the MGS repo, but we could use it throughout Nexus at least.)

Reply:
That's a good idea, but I will do it in a follow up PR (shortly).

Reply:
This ended up being smaller than I expected:
https://github.com/oxidecomputer/omicron/pull/8085/